### PR TITLE
Porting memory leak fixes from mainline repo

### PR DESCRIPTION
**Pull request by mderbaso-deepx**
### Description
Porting commits related to resource deallocation from mainline

### How Has This Been Tested?

ran empty simulation and one project-specific simulation

<br>
---

- [DeepX PR Guidelines](https://github.com/DeepX-inc/.github/blob/main/pull_request_guidelines.md) - for descriptions and metadata.
- [Google Code Review Guidelines](https://google.github.io/eng-practices/) - for [authors](https://google.github.io/eng-practices/review/developer/) and [reviewers](https://google.github.io/eng-practices/review/reviewer/).
- [Semantic Code Reviews](https://www.m31coding.com/blog/semantic-reviews.html) - Simple and direct comments without drama.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -39,23 +35,22 @@ RUN apt-get -qq update && \
     wget http://packages.osrfoundation.org/gazebo.key -O - | apt-key add - && \
     add-apt-repository ppa:kisak/kisak-mesa && \
     curl -sSL http://get.gazebosim.org | sh && \
-    update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-10 800 --slave /usr/bin/g++ g++ /usr/bin/g++-10 --slave /usr/bin/gcov gcov /usr/bin/gcov-10 && \
```
Comment (mderbaso-deepx) [inline Dockerfile:42]: we already have 11th version of compiler in image

File: Dockerfile [hunk 2]
```diff
@@ -39,23 +35,22 @@ RUN apt-get -qq update && \
...
     apt-get -qq update && \
     apt-get -qq upgrade -y --no-install-recommends \
-    libgz-cmake3-dev=3.2.* \
```
Comment (mderbaso-deepx) [inline Dockerfile:45]: in official docs they pin major version https://gazebosim.org/docs/harmonic/release-features#garden

File: Dockerfile (hunk 1 above)
Comment (mderbaso-deepx) [reply Dockerfile:42]: hm... it's not linking with g++10/11, it asks -fPIC, strange... Should investigate it, but later

Comment (mderbaso-deepx) [reply Dockerfile:42]: about linking: https://github.com/gazebosim/gz-rendering/pull/950

**Comment by mderbaso-deepx** [comment]
Maybe enable merge w/o squash for this repo?

File: ogre/src/OgreGpuRays.cc [hunk 3]
```diff
@@ -163,6 +163,9 @@ void OgreGpuRays::Init()
 //////////////////////////////////////////////////
 void OgreGpuRays::Destroy()
 {
+  if (!this->dataPtr->ogreCamera)
```
Comment (dmo-deepx) [inline ogre/src/OgreGpuRays.cc:166]: [question] I'm assuming the code kinda guarantees that if `this->dataPtr->ogreCamera` is `nullptr`, then the rest will also be `nullptr`, right?

File: ogre/src/OgreWideAngleCamera.cc [hunk 4]
```diff
@@ -180,6 +183,13 @@ void OgreWideAngleCamera::Destroy()
     }
   }
 
+  if (this->dataPtr->ogreCamera)
```
Comment (dmo-deepx) [inline ogre/src/OgreWideAngleCamera.cc:186]: This `if` is not needed since we have a check at the top of this method.

**Review by dmo-deepx — COMMENTED** [review]
Looks very minor but important.
Just that one question please.

File: ogre/src/OgreGpuRays.cc (hunk 3 above)
Comment (mderbaso-deepx) [reply ogre/src/OgreGpuRays.cc:166]: It's a backport changes from external source; same below

File: ogre/src/OgreWideAngleCamera.cc (hunk 4 above)
Comment (mderbaso-deepx) [reply ogre/src/OgreWideAngleCamera.cc:186]: correct, but it's a backport and some code above could overwrite this ptr to null probably

**Pull request merged**